### ISSUE TITLE
Switch to preferring smallest chain by default

### ISSUE THREAD

**Issue by mholt** [caddyserver/caddy#6035]
Currently, by default, Caddy (via CertMagic) will select the first certificate chain offered by the ACME server.

For Let's Encrypt, this is their cross-signed root chain, but the vast majority of clients accept Let's Encrypt's root by now.

We could reduce the size of the TLS Handshake by almost 33% if we switch to the smallest chain by default. CertMagic would still just use the first chain by default (basically letting the CA choose, which I think is reasonable). Caddy would just set `Smallest: true` in the `CertChainPreference` when invoking CertMagic.

Any objections?

More info: https://letsencrypt.org/certificates/

**Comment by francislavoie** [comment]
See https://letsencrypt.org/docs/certificate-compatibility/

Making the change to remove the root from the chain wouldn't impact any clients, because the ones that would have worked due to including the root (because of DST root cross-sign) no longer work as of 2021 because the DST root expired.

So I think I'm :+1: for making this change.

**Comment by mholt** [comment]
This forum post suggests, that for Android compatibility, we may want to wait until September 2024 (this year finally!): https://community.letsencrypt.org/t/production-chain-changes/150739

**Comment by francislavoie** [comment]
Would "smallest" cause us to use the X2 root (ECDSA) instead of RSA? That would harm compatibility if we do it by default too early.

**Comment by otbutz** [comment]
> Would "smallest" cause us to use the X2 root (ECDSA) instead of RSA?

The ECDSA chain seems to be opt-in per ACME account: https://community.letsencrypt.org/t/root-x2-alternate-chain-for-ecdsa-opt-in-accounts/202884

> That would harm compatibility if we do it by default too early.

They submitted the ECDSA root for inclusion in trust stores around April 2021: https://community.letsencrypt.org/t/isrg-root-x2-submitted-to-root-programs/149385

So while support should be widespread, X1 will work on many more devices.

**Comment by mholt** [comment]
Ok, thanks. I might leave this as-is for now. Maybe until closer to the end of the year.

**Comment by mholt** [comment]
Speaking of... https://community.letsencrypt.org/t/shortening-the-lets-encrypt-chain-of-trust/201580

> On 2024-02-08, we will stop providing the long chain by default, but clients can still be configured to request it.

**Comment by francislavoie** [comment]
Do we need to do anything, then? Are we using whatever default they give us?

**Comment by mholt** [comment]
Yeah, I guess so. The vast majority of users use Let's Encrypt anyway, and I don't think ZeroSSL offers multiple chains (?) and if other CAs do then they're already configuring them, so I don't know if there's any benefit of changing / complexifying our code at this point.

**Comment by alcohol** [comment]
> Speaking of... https://community.letsencrypt.org/t/shortening-the-lets-encrypt-chain-of-trust/201580
> 
> > On 2024-02-08, we will stop providing the long chain by default, but clients can still be configured to request it.

Is there a way to configure Caddy to request the long chain?

My client has some older handheld Android devices for which they have not had the time (nor finances) to upgrade them yet. Any months of prolonged use I can buy them they will be quite happy with. One of our certs got renewed this morning and ended up with the shorter chain, causing those devices to now be useless, until I can figure out how to automatically or manually retrieve longer chain certificates again. So I am very keen on any insights or information you might have.

**Comment by alcohol** [comment]
So what worked for me right now (quick and dirty perhaps..):

- Go to `/var/lib/caddy/.local/share/caddy/certificates/acme-v02.api.letsencrypt.org-directory/<domain.tld>`
- Get the currently used cert url `jq .issuer_data.url <domain>.<tld>.json`
- Check headers from `curl -I <url>` to get the alternate long chain cert location
  - parse long cert chain url from headers: `link: <long-cert-chain-url>;rel="alternate"`
- stop caddy
- Download long chain cert and overwrite existing cert: `curl <long-cert-chain-url> -o <domain>.<tld>.crt` 
- start caddy

**Comment by francislavoie** [comment]
See https://caddyserver.com/docs/caddyfile/directives/tls#acme `preferred_chains`. There's no "longest", but you can choose with the root common name.

**Comment by alcohol** [comment]
Would the following as a global configuration be sufficient then:

```
{
  tls {
    issuer acme {
      preferred_chains {
        root_common_name "DST Root CA X3"
      }
    }
  }
}

**Comment by alcohol** [comment]
Oh wait, https://caddyserver.com/docs/caddyfile/options#preferred-chains does allow it on a global level.

**Comment by francislavoie** [comment]
Yep - or you can use [`cert_issuer`](https://caddyserver.com/docs/caddyfile/options#cert-issuer) which is the same as `tls > issuer`, but global

**Comment by alcohol** [comment]
Thanks for the quick responses! Helped me a lot :partying_face:

**Comment by AndreKR** [comment]
So I would use...

Globally for both issuers:
```
{
  preferred_chains {
    root_common_name "DST Root CA X3"
  }
}
```

Globally for one issuer:
```
{
  cert_issuer acme {
    preferred_chains {
      root_common_name "DST Root CA X3"
    }
  }
}
```

And for a single site:
```
my-site.com {
  tls {
    issuer acme {
      preferred_chains {
        root_common_name "DST Root CA X3"
      }
    }
    issuer zerossl {
      preferred_chains {
        root_common_name "DST Root CA X3"
      }
    }
  }
}
```

Correct?

On a site level I have to configure it for both issuers separately, right?

Afterwards it seems I have to delete the certificate from the store and then restart (not just reload) Caddy for the changes to take effect.

**Comment by mholt** [comment]
Yes I think that is right, without having verified all the names and such.

The reason a restart is required is because Caddy persists valid certs through reloads to keep things lightweight and efficient. Basically changes affect the next obtain/renew they don't change out all the certs right away.

**Comment by AndreKR** [comment]
The name should be correct, I copy&pasted this from my working config.

By the way, are you aware that the default chain prevents devices running Android < 7.1.1 from visiting sites that are hosted by Caddy? This seems like something that could also have been by accident.

**Comment by alcohol** [comment]
They are very much aware, several community/blog/news posts were made about this. E.g. https://community.letsencrypt.org/t/shortening-the-lets-encrypt-chain-of-trust/201580 - but there were other posts as well.

**Comment by AndreKR** [comment]
Hm, I see. Looks like this is a short-lived solution. I guess I need to switch to ZeroSSL or see if I can distribute a trust root as part of my app.

**Comment by mholt** [comment]
> By the way, are you aware that the default chain prevents devices running Android < 7.1.1 from visiting sites that are hosted by Caddy?

If supporting EOL'ed 8-year-old devices is important to your site, then I'd recommend configuring accordingly; we also don't support many operating systems that are about that old (for example, Windows XP was EOL'ed not long before that), etc etc. But you can configure your site to support old clients if needed.

The cert chain is expiring soon so they won't work at that point anyway.

**Comment by otbutz** [comment]
> in a way that also has a lifespan with a known EOL date?

https://letsencrypt.org/2023/07/10/cross-sign-expiration.html

tl;dr: LetsEncrypt's cross-signed intermediate certificates expire on 30 September 2024

**Comment by AndreKR** [comment]
I think you can use ZeroSSL, their root is trusted in all Android versions.
